pkg/domain/infra/abi/quadlet.go

```diff
@@ -0,0 +1,699 @@
+	}
+
+	finalPath := filepath.Join(installDir, filepath.Base(filepath.Clean(path)))
+	if destName != "" {
+		finalPath = filepath.Join(installDir, destName)
+	}
+
+	// Second, validate that the dest path does NOT exist.
+	// TODO: Overwrite flag?
+	if _, err := os.Stat(finalPath); err == nil {
+		return "", fmt.Errorf("a Quadlet with name %s already exists, refusing to overwrite", filepath.Base(finalPath))
+	}
+
+	// Validate extension is valid
+	if validateQuadletFile && !systemdquadlet.IsExtSupported(finalPath) {
+		return "", fmt.Errorf("%q is not a supported Quadlet file type", filepath.Ext(finalPath))
+	}
+
+	// Move the file in
+	contents, err := os.ReadFile(path)
+	if err != nil {
+		return "", fmt.Errorf("reading source file %q: %w", path, err)
+	}
+	if err := os.WriteFile(finalPath, contents, 0644); err != nil {
+		return "", fmt.Errorf("writing Quadlet %q to %q: %w", path, finalPath, err)
+	}
+
+	// TODO: It would be nice to do single-file validation here, and remove the file if it fails.
+	if !validateQuadletFile {
+		err := util.AppendStringToFile(filepath.Join(installDir, assetFile), filepath.Base(filepath.Clean(path)))
+		if err != nil {
+			return "", fmt.Errorf("error while writing non-quadlet filename: %w", err)
+		}
+	}
+	return finalPath, nil
+}
+
+// buildAppMap scans the given directory for files that start with '.'
+// and end with '.app', reads their contents (one filename per line), and
+// returns a map where each filename maps to the .app file that contains it.
+// Also returns a map where each `.app` points to a slice of strings containing
+// all the files in that `.app`.
+func buildAppMap(dir string) (map[string]string, map[string][]string, error) {
+	reverseMap := make(map[string]string)
+	appMap := make(map[string][]string)
+
+	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() {
+			name := info.Name()
+			if strings.HasPrefix(name, ".") && strings.HasSuffix(name, ".app") {
+				file, err := os.Open(path)
+				if err != nil {
+					return err
+				}
+				defer file.Close()
+
+				scanner := bufio.NewScanner(file)
+				for scanner.Scan() {
+					filename := strings.TrimSpace(scanner.Text())
+					if filename != "" {
+						reverseMap[filename] = name
+						appMap[name] = append(appMap[name], filename)
+					}
+				}
+				if err := scanner.Err(); err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	return reverseMap, appMap, nil
+}
+
+// Get the paths of all quadlets available to the current user
+func getAllQuadletPaths() []string {
+	var quadletPaths []string
+	quadletDirs := systemdquadlet.GetUnitDirs(rootless.IsRootless())
+	for _, dir := range quadletDirs {
+		dents, err := os.ReadDir(dir)
+		if err != nil {
+			// This is perfectly normal, some quadlet directories aren't created by the package
+			logrus.Infof("Cannot list Quadlet directory %s: %v", dir, err)
```

Comment:
if it is normal then there is no reason to confuse users with errors if they run with log level info?
I would say ErrNotExist should be ignored but other errors definly should be logged then on warning level.